### PR TITLE
Refactor cluster state to avoid checking NodeType everywhere and redu…

### PR DESCRIPTION
**Pull request by baldersheim**
…ce code duplication.

@vekterli PR
Should be pure refactoring

### CODE REVIEW

File: vdslib/src/main/java/com/yahoo/vdslib/state/ClusterState.java [hunk 1]
```diff
@@ -18,25 +19,183 @@
...
+        }
+
+        int getMaxIndex() { return maxIndex; }
+
+        NodeState getNodeState(int index) {
+            NodeState ns = nodeStates.get(index);
+            if (ns != null) return ns;
+            return (index >= getMaxIndex() || ! upNodes.get(index))
+                    ? new NodeState(type, State.DOWN)
+                    : new NodeState(type, State.UP);
```
Comment (vekterli) [inline vdslib/src/main/java/com/yahoo/vdslib/state/ClusterState.java:59]: More fuel for the "node states should be immutable" fire... 🔥

**Review by vekterli — APPROVED** [review]
:+1: a few minor comments. `memory_overhead--;`

File: vdslib/src/main/java/com/yahoo/vdslib/state/ClusterState.java [hunk 2]
```diff
@@ -18,25 +19,183 @@
...
     private static final NodeState DEFAULT_DISTRIBUTOR_UP_NODE_STATE = new NodeState(NodeType.DISTRIBUTOR, State.UP);
+    private static final NodeState DEFAULT_STORAGE_DOWN_NODE_STATE = new NodeState(NodeType.STORAGE, State.DOWN);
+    private static final NodeState DEFAULT_DISTRIBUTOR_DOWN_NODE_STATE = new NodeState(NodeType.DISTRIBUTOR, State.DOWN);
 
-    private int version = 0;
-    private State state = State.DOWN;
-    // nodeStates maps each of the non-up nodes that have an index <= the node count for its type.
-    private Map<Node, NodeState> nodeStates = new TreeMap<>();
+    /**
+     * Maintains a bitset where all non-down dows have a bit set. All nodes that differs from defaultUp
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vdslib/src/main/java/com/yahoo/vdslib/state/ClusterState.java]: Fixed

File: vdslib/src/main/java/com/yahoo/vdslib/state/ClusterState.java [hunk 3]
```diff
@@ -18,25 +19,183 @@
...
+            if (node.getType() != type) {
+                throw new IllegalArgumentException("NodeType '" + node.getType() + "' differs from '" + type + "'");
+            }
+        }
+
+        void setNodeState(Node node, NodeState ns) {
+            validateInput(node, ns);
+            int index = node.getIndex();
+            if (index >= maxIndex) {
+                maxIndex = index + 1;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vdslib/src/main/java/com/yahoo/vdslib/state/ClusterState.java]: Agree, fixed.

File: vdslib/src/main/java/com/yahoo/vdslib/state/NodeState.java [hunk 4]
```diff
@@ -51,6 +51,7 @@ public boolean above(NodeState other) {
     public boolean equals(Object o) {
         if (!(o instanceof NodeState)) { return false; }
         NodeState ns = (NodeState) o;
+        // Note that 'description' is not considered. Intentional ?
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vdslib/src/main/java/com/yahoo/vdslib/state/NodeState.java]: Updated comments.

File: vdslib/src/main/java/com/yahoo/vdslib/state/ClusterState.java (hunk 1 above)
Comment (baldersheim) [reply vdslib/src/main/java/com/yahoo/vdslib/state/ClusterState.java:59]: This breaks down if someone does getNodeState(node_that_is_up).modify_some_important_field.

**Pull request merged**